### PR TITLE
Feature/issue16 autofac lazy func

### PR DESCRIPTION
**Pull request by cvaky**
Added support for Lazy<T> and Func<T>, Owned<T> in constructor

### CODE REVIEW

File: src/AgentMulder.Containers.Autofac/Patterns/NavigateOwned.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+using JetBrains.ReSharper.Psi.CSharp.Tree;
+
+namespace AgentMulder.Containers.Autofac.Patterns
+{
+    [Export(typeof(INavigationProvider))]
+    public class NavigateOwned : NavigationProvider
+    {
+        public override bool CheckOwned(ICSharpParameterDeclaration paramNode)
+        {
+            var result = paramNode?.Type?.GetScalarType()?.GetClrName().ShortName.ToLower() == "owned";
```
Comment (mareklinka) [inline src/AgentMulder.Containers.Autofac/Patterns/NavigateOwned.cs]: This is not an optimal way to check for it. If you can't compare type instances (which probably won't be possible), at least check the full type name.

File: src/AgentMulder.ReSharper.Domain/Patterns/INavigationProvider.cs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+using JetBrains.ReSharper.Psi;
+using JetBrains.ReSharper.Psi.CSharp.Tree;
+
+namespace AgentMulder.ReSharper.Domain.Patterns
+{
+    public interface INavigationProvider
```
Comment (mareklinka) [inline src/AgentMulder.ReSharper.Domain/Patterns/INavigationProvider.cs:7]: Name feels too generic. I'd say this is something that provides navigation services. Can we think of something more specific?

File: src/AgentMulder.ReSharper.Plugin/Navigation/GotoInstantiationNavigationProvider.cs [hunk 3]
```diff
@@ -24,9 +27,26 @@ namespace AgentMulder.ReSharper.Plugin.Navigation
     [ContextNavigationProvider]
     public class GotoInstantiationNavigationProvider : INavigateFromHereProvider
     {
+        [Import(typeof(INavigationProvider))]
+        private INavigationProvider navigationProvider;
```
Comment (mareklinka) [inline src/AgentMulder.ReSharper.Plugin/Navigation/GotoInstantiationNavigationProvider.cs]: There is a possibility to have more providers, right? Maybe this should import a collection?

**Review by mareklinka — CHANGES_REQUESTED** [review]
Some improvements should be made here, please check the individual comments.

File: src/AgentMulder.ReSharper.Domain/Patterns/INavigationProvider.cs (hunk 2 above)
Comment (cvaky) [reply src/AgentMulder.ReSharper.Domain/Patterns/INavigationProvider.cs:7]: Not sure how could I name it. What would you say for IGotoInstantiationNavigationProviderExtensions or INavigationProviderExtensions ? In the future, there is a possibility to use this interface in GoToRegisteredComponentsNavigationProvider.cs

File: src/AgentMulder.ReSharper.Plugin/Navigation/GotoInstantiationNavigationProvider.cs (hunk 3 above)
Comment (cvaky) [reply src/AgentMulder.ReSharper.Plugin/Navigation/GotoInstantiationNavigationProvider.cs]: Changed to IEnumerable

**Review by mareklinka — APPROVED** [review]

**Pull request merged**